### ISSUE TITLE
newline-per-chained-call not working as expected

### ISSUE THREAD

**Issue by Sawtaytoes** [eslint/eslint#12970]
**What rule do you want to change?**
`newline-per-chained-call`

**Does this change cause the rule to produce more or fewer warnings?**
More.

**How will the change be implemented? (New option, new default behavior, etc.)?**
A new option.

**Please provide some example code that this change will affect:**

_**NOTE:** D3 example copied straight from the docs._
```js
// ignoreChainWithDepth: 1
let d3

// Currently allowed.
d3.select("body")
.selectAll("p")
.data([4, 8, 15, 16, 23, 42 ])

// What I'd expect to see.
d3
.select("body")
.selectAll("p")
.data([4, 8, 15, 16, 23, 42 ])

let someReallyLongArrayName = []

// Currently allowed.
someReallyLongArrayName.map(({
  foo,
  bar,
}) => ({
  foo,
  bar,
}))
.filter(Boolean)

// What I'd expect to see.
someReallyLongArrayName 
.map(({
  foo,
  bar,
}) => ({
  foo,
  bar,
}))
.filter(Boolean)
```

**What does the rule currently do for this code?**
With the `ignoreChainWithDepth` value, there's no way to say "if the chain depth is no longer ignored, _all_ values need newlines".

At the point that this rule is triggered, instead of allowing only the first chained prop to stay at the top, all props would be on a new line.

**What will the rule do after it's changed?**
It will look to see if the chain depth has been passed and if so, enforce also dropping the first value for consistency. It's strange to have one chain not being dropped, but the rest having to be dropped.

**Are you willing to submit a pull request to implement this change?**
Yes.

**Comment by mdjermanovic** [comment]
It isn't only the first, the rule allows first `ignoreChainWithDepth` elements in the chain to stay on the same line (or be in any place), so this is also valid with the default `ignoreChainWithDepth: 2`

```js
/*eslint newline-per-chained-call: ["error", { "ignoreChainWithDepth": 2 }]*/

// valid
a.b().c()
  .d();
```

I believe this is actually a bug introduced in a fix for another issue because it seems that the initial version of this rule didn't work like this. It used to require newlines before all elements of the chain if the total depth is greater than `ignoreChainWithDepth`.

We even have a wrong  **incorrect**  example in the documentation now:

```js
/*eslint newline-per-chained-call: ["error", { "ignoreChainWithDepth": 2 }]*/

_
  .chain({}).map(foo)
  .filter(bar);
```

This example is at the moment *correct* by this rule.

However, fixing this would be a big change in the default behavior now, so I support this request to add an option with a default value that would retain the current behavior. 

In my opinion, while technically an enhancement, this should be accepted as a bug fix.

**Comment by mdjermanovic** [comment]
@Sawtaytoes to avoid confusion, this example should be still *valid* with default settings, because the depth is 2:

```js
// Currently allowed.
someReallyLongArrayName.map(({
  foo,
  bar,
}) => ({
  foo,
  bar,
}))
.filter(Boolean)
```

**Comment by Sawtaytoes** [comment]
I understand. You're saying the behavior was modified and instead of changing it back, we'll keep it as-is and add new functionality to allow dropping the `.` when there are more than 1 `.` chain or more than X `.` chains right?

**Comment by mdjermanovic** [comment]
> I understand. You're saying the behavior was modified and instead of changing it back, we'll keep it as-is and add new functionality to allow dropping the `.` when there are more than 1 dot chain or more than X dot chains right?

That's the idea, but let's wait for more opinions from other team members before doing any work.

I think this behavior was modified in #5365, probably accidentally.

In `v2.2.0` it used to work like this:

```js
/*eslint newline-per-chained-call: [2, { "ignoreChainWithDepth": 2 }]*/

// invalid
a.b().c().d();

// invalid
a.b().c()
  .d();

// invalid
a.b()
  .c()
  .d();  

// valid
a
  .b()
  .c()
  .d();    
```

In the actual `v6.8.0` it's:

```js
/*eslint newline-per-chained-call: [2, { "ignoreChainWithDepth": 2 }]*/

// invalid
a.b().c().d();

// valid
a.b().c()
  .d();

// valid
a.b()
  .c()
  .d();  

// valid
a
  .b()
  .c()
  .d();  
```

A solution might be to add an option that accepts a number of leading elements to ignore (defaults to the number specified in `ignoreChainWithDepth`, which retains the current behavior).

**Comment by Sawtaytoes** [comment]
Yes, the v2 version is what I'd expect to see. Thanks for looking into this!

**Comment by kaicataldo** [comment]
Given how old this option is, I'm not sure it's worth changing it at this point. This could potentially add a lot of pain to the upgrade process.

**Comment by Sawtaytoes** [comment]
I understand. I'll create a PR to add a new prop to do the functionality I'd expect from v2.

**Comment by mdjermanovic** [comment]
I was thinking about a non-breaking change like this:

```js
/*eslint newline-per-chained-call: ["error", 
  { 
    "ignoreChainWithDepth": 4
    "allowLeading": 2
  }
]*/

// valid
obj.a().b().c().d();

// invalid
obj.a().b().c().d().e();

// invalid. this would be valid without allowLeading
obj.a().b().c().d()
  .e();

// invalid. this would be valid without allowLeading
obj.a().b().c()
  .d()
  .e();

// valid. the two invalid examples above would be autofixed to this
obj.a().b()
  .c()
  .d()
  .e();

// also valid
obj.a()
  .b()
  .c()
  .d()
  .e();

// also valid
obj
  .a()
  .b()
  .c()
  .d()
  .e();
```

`allowLeading` is a new option. There could be a better name, though.

If it isn't set, the behavior is the same as the actual behavior.

`"allowLeading": 0` restores behavior from `v2` (only the last example would be valid).

**Comment by Sawtaytoes** [comment]
I think I get the difference in functionality here. Took me a while to re-read and re-understand.

I want `ignoreChainWithTotalDepth` whereas right now, it's `ignoreChainWithAttachedDepth`.

Sadly, I can't really make it one or the other at this point with the way options are handled and the current behavior is confusing as it does not specify "total" or "attached" in the prop name.

I understand the `allowLeading` variable name, but that's probably confusing as well since it's not just leading, it's more like "this is the maximum total length your chain can be before _everything_ needs to be dropped.

By specifying "leading", we don't want a situation like this:

```js
/*eslint newline-per-chained-call: ["error", 
  { 
    "ignoreChainWithDepth": 4
    "allowLeading": 2
  }
]*/

obj.a().b()
  .c().d()
  .e();
```

So I'm going to go for `maxTotalChainDepth`. That should make it more obvious to people this is the most items you can have in a chain of commands before all have to be dropped. Default value of `Infinity` to retain existing behavior.

**Comment by Sawtaytoes** [comment]
Here's a preview:

![image](https://user-images.githubusercontent.com/3948069/75954910-64e56e00-5e7a-11ea-962d-e261ff7a9056.png)

I'll have to separate it so only the `CallExpression` types are underlined and each one is separately identified in `context.report`, but functionality-wise, I figured out how to identify total chain length.

The code right now is looking at each `CallExpression` separately. I'm looking at for `ExpressionStatement` in the `parent` to find out which `CallExpression` is the final one:

![image](https://user-images.githubusercontent.com/3948069/75955031-b42b9e80-5e7a-11ea-883f-8ca9b9564f4b.png)

Thing is, the old code also looked for `MemberExpression` as well as `CallExpression`. I think, based on the rule name, it makes sense to only look for `CallExpression`, but here's what happens then:

![image](https://user-images.githubusercontent.com/3948069/75955251-1389ae80-5e7b-11ea-88e9-2b73eff2accc.png)

Now the `MemberExpression` `.length` has the `ExpressionStatement` parent, and I can't enforce max total chain depth. It doesn't care how many members you have, it only enforces newlines on consecutive `CallExpression` depth:

![image](https://user-images.githubusercontent.com/3948069/75955371-53509600-5e7b-11ea-9f6c-aed6c89276d9.png)

This is something that needs to be re-examined because I think some of this is clearly unintended behavior.

Now I'm looking into pairing `CallExpression` and `MemberExpression` as another option. I'm really thinking the current rule needs to be renamed to `newline-per-chained-member` as it's more generic.

My goal now is figuring out if I can maintain the old functionality along with this new prop and include `MemberExpression` as a potential 3rd option without the whole thing breaking down.

**Comment by Sawtaytoes** [comment]
This is what I ended up doing:

```js
            "CallExpression:exit"(node) {
                if (!node.callee || node.callee.type !== "MemberExpression") {
                    return;
                }

                if (isUsingMaxTotalChainDepth) {
                    validateMaxTotalChainDepth(node);
                } else {
                    validateCallExpressionIgnoreDepth(node);
                }
            },

            "MemberExpression:exit"(node) {
                if (!isUsingMaxTotalChainDepth) {
                    return;
                }

                validateMaxTotalChainDepth(node);
            }
```

PR coming soon.

**Comment by mdjermanovic** [comment]
The idea with the new `number` option was to provide more flexibility with, I believe, about the same added complexity in the rule's code.

For example, someone might want to allow only the first `.method()` to be on the same line as the object when the total depth exceeds `ignoreChainWithDepth`:

```js
/*eslint newline-per-chained-call: ["error", 
  { 
    "ignoreChainWithDepth": 4
    "allowLeading": 1
  }
]*/

obj.a()
  .b()
  .c()
  .d()
  .e();
```

```js
/*eslint newline-per-chained-call: ["error", 
  { 
    "ignoreChainWithDepth": 4
    "allowLeading": 0
  }
]*/

obj
  .a()
  .b()
  .c()
  .d()
  .e();
```

However, if it doesn't add much value, and especially if it wouldn't be clear what's the purpose of having two `number` options, then we should just add a `boolean` option:

```js
/*eslint newline-per-chained-call: ["error", 
  { 
    "ignoreChainWithDepth": 4
    "all": true
  }
]*/

obj
  .a()
  .b()
  .c()
  .d()
  .e();
```

**Comment by Sawtaytoes** [comment]
@mdjermanovic That's a fantastic solution!

Have one number prop as before, but have the capability of changing that functionality with a `boolean` or even a descriptor `string` if we needed more versions.

I'll add that into the PR.

**Comment by mdjermanovic** [comment]
> Have one number prop as before, but have the capability of changing that functionality with a `boolean` or even a descriptor `string` if we needed more versions.

I'm not sure that I understand the `string` part, what string values could it accept?

**Comment by Sawtaytoes** [comment]
Since I don't know what people are going to be adding in the future, here's what I've got so far. PR's not updated yet with my latest changes.

```js
depthCalculationStyle: {
    type: "string",
    enum: ["all", "perLine"],
    default: "perLine"
},
ignoreChainWithDepth: {
    type: "integer",
    minimum: 1
},
includeBrackets: {
    type: "boolean",
    default: true
},
includeMethodCalls: {
    type: "boolean",
    default: true
},
includeProperties: {
    type: "boolean",
    default: false
}
```

**Comment by mdjermanovic** [comment]
@Sawtaytoes I think that we should split this into several different proposals, i.e. make a separate issue for each of these enhancements.

This looks like to me like three unrelated enhancements:

1. In my understanding, only this is related to the proposal from the description of this issue:

```js
depthCalculationStyle: {
    type: "string",
    enum: ["all", "trailingMembers"],
    default: "trailingMembers"
},
```

2. This is another enhancement (it could be actually a bug fix):

```js
includeBracketedProperties: {
    type: "boolean",
    default: true
},
```

3. This is another enhancement:

```js
includeMethodCalls: {
    type: "boolean",
    default: true
},
includeProperties: {
    type: "boolean",
    default: false
}
```

In order to be accepted, an enhancement needs a champion from the team + 3 :+1: from other team members as outlined [here](https://eslint.org/docs/maintainer-guide/issues#accepting-issues), so bundling several different proposals into one makes it difficult to evaluate and eventually accept the issue.

I personally think that 1. and 2. are quite reasonable (though I'm a bit confused by the design for 1. at the moment), but I'm not sure that we should implement 3.

**Comment by Sawtaytoes** [comment]
@mdjermanovic I'll figure out a way to split this into multiple issues so I can get + 3 easily. I've already spent over 20 hours working on this fix and breaking it up will be harder mainly because I'm constantly force pushing to retain the

**Comment by mdjermanovic** [comment]
> I'm constantly force-pushing to retain the commit message part of the PR. Is that PR check only for the PR title or the commits themselves?

If PR has only 1 commit, then the commit message is checked. If it has multiple commits (i.e. if you are adding commits instead of force-pushing one), then only the PR title matters.

**Comment by mdjermanovic** [comment]
> ```js
> depthCalculationStyle: {
>     type: "string",
>     enum: ["all", "trailingMembers"],
>     default: "trailingMembers"
> },
> ```

I find the `depthCalculationStyle` name confusing, because the depth is always the same:

```js
a.b(); // 1

a.b().c(); // 2

a.b().c().d(); // 3
``` 
User could expect that this option changes the meaning of "depth" and that it consequently affects how the rule determines whether particular chain has a depth greater than the value set in `ignoreChainWithDepth`.

Depth calculation style might be an implementation detail, but the observable change for users is only whether the rule will require newlines for all members of the chain once it has already calculated the depth and determined whether it exceeds `ignoreChainWithDepth` regardless of this option?

Thoughts about something like this:

```js
requireNewlineFor: {
    type: "string",
    enum: ["allMembers", "trailingMembers"],
    default: "trailingMembers"
},
```
